java/pattern/schema/Rule.java

```diff
@@ -157,6 +160,15 @@ private static void validateThen(String label, @Nullable Conjunction<? extends P
                 throw TypeQLException.of(INVALID_RULE_THEN_VARIABLES.message(label));
             }
         }
+
+        // Roles must be explicit
+        if (then.relation().isPresent() && !then.relation().get().players().stream()
+                .map(player -> {
+                    return player.roleType().isPresent();
+                })
```

Comment:
for a single statement you don't need `{}`, you can just do it in one line: `.map(player -> player.roleType().isPresent()`